### ISSUE TITLE
Check CqlEvaluatorSampleApp & Integrate CQL Library as demo app into FHIR core

### ISSUE THREAD

**Issue by f-odhiambo** [opensrp/fhircore#177]
This is a simple app that creates a cql-evaluator and runs the WHO ANC Recommendation A2 content. It was created using Android Studio. Link to [repo](https://github.com/DBCG/CqlEvaluatorSampleApp)

Checks 
- [x] Check the structure of the FHIR resources being loaded from the assets folder 
- [x] Test functionality of loading and execution of CQL on Android client
- [x] Check whether $measure_evaluate is being run locally and the related functions on the sample app
- [x] Document the finding on this issue/ticket
 

Sample Screenshots

<img src="https://user-images.githubusercontent.com/4540684/125922970-0d0be216-33e8-49b2-8f62-a6af107a1815.png" width="200" height="400">.  <img src="https://user-images.githubusercontent.com/4540684/125922979-b0a4332c-4d06-4645-8e61-594a7bd7db00.png" width="200" height="400">

Integration
- [x] Integrate it into FHIR core as a demo app for mADX  with some enhancements which include 
- [ ] Removed reference FHIR resources from Assets folder on the app and instead sync them from HAPI FHIR Server
- [x] List Patients from the server and not assets folder
- [ ] Add screen for Selection of patient and definition to generate the operation definition 

After this we  proceed to
- [ ] #404 
- [ ] #405 
- [ ] Run $measure_evaluate on server side and then later locally on Android client using the Workflow API

**Comment by f-odhiambo** [comment]
Bug in the demo app  : Seems like there are some libraries that are not syncing anymore and closing the app to terminate before launch
1. Could not find org.opencds.cqf.cql:evaluator:1.2.0-SNAPSHOT.
2. Could not find org.opencds.cqf.cql:evaluator.builder:1.2.0-SNAPSHOT.
3. Could not find org.opencds.cqf.cql:evaluator.dagger:1.2.0-SNAPSHOT.

Bug reported here https://github.com/DBCG/CqlEvaluatorSampleApp/issues/1#issue-909546106

**Comment by f-odhiambo** [comment]
Updated the libs and now working fine

**Comment by rkodev** [comment]
### Summary of the Sample CQL Application

The app is able to evaluate CQL expressions written in `text/cql`. 
The sample app demonstrates the ability to load a [FHIR Library ](https://www.hl7.org/fhir/library.html) and [Value Sets](https://www.hl7.org/fhir/valueset.html) thats are later used to evaluate the Libraries parameter output against a [Patient Resource](https://www.hl7.org/fhir/patient.html) as an input parameter.

As of the time of this post, neither `$mesare_evaluate` nor `$apply` operations are capable with the version of this app.

It should be possible to integrate the library as it is to the fhir-core application but for some major concerns with the duration it takes to run a single evaluation operation with CQL.

**Comment by pld** [comment]
1. How we will evaluate Measure(s) to get MeasureReport(s)
  - should any of the work we do towards this issue go upstream into the CQL libs?
2. How and what parts of this will go into the “engine” section and what into the “chw anc” app section?
  - and are there other custom pieces not in either of those?

**Comment by MuhammadSalman-7214** [comment]
@f-odhiambo so we will have different flow for mADX and Covax application going forward, like we would have a screen after sign up that can have two buttons. One for Covax application, and one for mADX application. User can press button and go to the specific flow of that application like for mADX, mADX flow will run and for COVAX, COVAX flow will run, right?

**Comment by f-odhiambo** [comment]
Not really. the Single APK application will be defined via a configuration that will check the credentials of the user and the load the appropriate workflow either mADX or Covax

**Comment by MuhammadSalman-7214** [comment]
okay, @f-odhiambo what actually I want to ask is whether this mADX application will be running as a separate application from fhir core, or it will be used as a same application within fhir core?

**Comment by pld** [comment]
Ie the first task under integration needs more definition, before we begin on it, but the others are unblocked

**Comment by MuhammadSalman-7214** [comment]
> 1. How we will evaluate Measure(s) to get MeasureReport(s)
>    
>    * should any of the work we do towards this issue go upstream into the CQL libs?
> 2. How and what parts of this will go into the “engine” section and what into the “chw anc” app section?
>    
>    * and are there other custom pieces not in either of those?

@f-odhiambo can you please answer this one?

**Comment by MuhammadSalman-7214** [comment]
> > 1. How we will evaluate Measure(s) to get MeasureReport(s)
> >    
> >    * should any of the work we do towards this issue go upstream into the CQL libs?
> > 2. How and what parts of this will go into the “engine” section and what into the “chw anc” app section?
> >    
> >    * and are there other custom pieces not in either of those?
> 
> @f-odhiambo can you please answer this one?

as far as I can answer, currently it will be independent of both covax and chw application

**Comment by MuhammadSalman-7214** [comment]
> Ie the first task under integration needs more definition, before we begin on it, but the others are unblocked

yes have a call with @f-odhiambo regarding this, and currently we decided to proceed with cql as a separate application than fhir core

**Comment by MuhammadSalman-7214** [comment]
There can be two approaches to integrate CQL evaluator application into Fhir Core:
1. Either we integrate it as a separate application from fhir core, which means we would not be able to access code from fhir core in cql application. 
2. We can integrate cql application as a module into fhir core, and then include this cql module into fhir core, but for this approach as well, the code for cql stuff would be written in cql module, but here we will be able to access code from fhir core.

**Comment by pld** [comment]
Sounds like we need to go w/(2) since the value of integration is being able to call this functionality within FHIR Core

I also do not think we should include any of the UI components. My understanding is that the goal of the integration is so we can implement #404 and #405, not have a CQL demo

**Comment by pld** [comment]
> > Ie the first task under integration needs more definition, before we begin on it, but the others are unblocked
> 
> yes have a call with @f-odhiambo regarding this, and currently we decided to proceed with cql as a separate application than fhir core

Sounds good, @ekigamba can you join this as well? After the meeting please post the plan here so I can take a look, I'm happy to join as well, but definitely proceed w/out me and I can catch up, I don't want to slow you down!

**Comment by MuhammadSalman-7214** [comment]
> > > Ie the first task under integration needs more definition, before we begin on it, but the others are unblocked
> > 
> > 
> > yes have a call with @f-odhiambo regarding this, and currently we decided to proceed with cql as a separate application than fhir core
> 
> Sounds good, @ekigamba can you join this as well? After the meeting please post the plan here so I can take a look, I'm happy to join as well, but definitely proceed w/out me and I can catch up, I don't want to slow you down!

@pld if we are going with second approach, then this is now no longer valid

**Comment by pld** [comment]
True, we can rewrite the title and description of the issue to match this.

My assumption is that there's code in this app that's useful for #404 and #405, maybe it'd be more useful to work on one of those issues directly and treat this as a sub-issue of those to pull in common CQL functionality.

@f-odhiambo am I missing something here?

**Comment by ekigamba** [comment]
> > > Ie the first task under integration needs more definition, before we begin on it, but the others are unblocked
> > 
> > 
> > yes have a call with @f-odhiambo regarding this, and currently we decided to proceed with cql as a separate application than fhir core
> 
> Sounds good, @ekigamba can you join this as well? After the meeting please post the plan here so I can take a look, I'm happy to join as well, but definitely proceed w/out me and I can catch up, I don't want to slow you down!

Sure @pld

**Comment by f-odhiambo** [comment]
> True, we can rewrite the title and description of the issue to match this.
> 
> My assumption is that there's code in this app that's useful for #404 and #405, maybe it'd be more useful to work on one of those issues directly and treat this as a sub-issue of those to pull in common CQL functionality.
> 
> @f-odhiambo am I missing something here?

@pld We will need to extend the work Ronald was working on for the search use cases to achieve this. Search will need to be extended to work for either name (First name, Last Name, Family name), IDs (UUID, Logical ID, Identifiers) and resources (Encounters, Observations, Practitioners etc) for issues #404 & #405 

For this piece of work, I think we are looking at the base code, resources to achieve measure reporting by using the Parameter bundle resource it generates, and later to be consumed by HAPI for $measure_evalute and generate a measure report

**Comment by f-odhiambo** [comment]
> Sounds like we need to go w/(2) since the value of integration is being able to call this functionality within FHIR Core
> 
> I also do not think we should include any of the UI components. My understanding is that the goal of the integration is so we can implement #404 and #405, not have a CQL demo

@pld for the mADX work we need to achieve the following 
1. Log in to the app 
2. View a list of patient 
3. Search (Patient /Encounter / Observations)
4. Define the _measure_ for the indicator (for POC we shall go with Indicator i.e. Percentage of pregnant women with first ANC contact in the first trimester (before 1 week of gestation) ) we need to report 
5. Run $measure_evalute 
6. Show measure report

**Comment by f-odhiambo** [comment]
So we skip moving from Evaluation page directly to map the floe above

**Comment by pld** [comment]
To clarify, I meant the UI of this specific demo app is not required.

To achieve that we can use the ANC app NOT a separate CQL specific demo app.

**Comment by f-odhiambo** [comment]
Copy that. All Clear

**Comment by MuhammadSalman-7214** [comment]
So based on @pld comments that we don't need the UI of the demo application, the approach finalized with @f-odhiambo and @ekigamba is as follows:

We will not import the cql demo application as a module into fhir core, instead we will copy some of the code and dependencies directly from reference application to fhir core and then use it to achieve our use cases within fhir core directly

**Comment by MuhammadSalman-7214** [comment]
**Update:**
1. Cql Evaluator reference application has been integrated into fhir core
2. We have uploaded all Library Resources and Plan Definition Resource to server, and they are being used from server instead of assets folder, but we still need to upload valueset and test bundle files to server. Currently both valueset and test bundle files are being used from asset folder
3. Though we have uploaded Plan Definition resource file to server, but currently there isn't any usage of the plan definition in the codebase(code of reference application that has been integrated to fhir core)

@f-odhiambo

**Comment by pld** [comment]
awesome work! is it ready for a PR yet?

**Comment by MuhammadSalman-7214** [comment]
> awesome work! is it ready for a PR yet?

yes almost

**Comment by f-odhiambo** [comment]
Canonical Types are not being handled yet in FHIR SDK tracked under issues https://github.com/google/android-fhir/issues/694. Until this is resolved we will not be able to sync value sets and test JSON from server into the local SQLite DB

**Comment by f-odhiambo** [comment]
**Acceptance Criteria** 
- [ ] Checkout  `feature/integrate-cql-evaluation-application`
- [ ] Make sure that we are loading resources from the HAPI server and not the assets folder. Some resources might not be in use/required  but confirm 
- [ ] Load  Patient resources from HAPI. NB. You will need to edit the patient resource and add the address given that the current app filters/syncs patients by location i.e. Nairobi 
```
 "address": [ {
        "city": "Nairobi",
        "country": "Kenya"
      } ]
```
- [ ] Add an Evaluate CQL button from the Patient profile
- [ ] Add a text area/label to view results i.e. Parameter resource payload. NB. This should only be for this target sample patient and not the rest of the other patients 
- [ ] Will need to publish this as a module on the Single APK Approach. Martin or Elly can guide

**Comment by f-odhiambo** [comment]
@kimigxfoxy Also take note that #414 is now complete and that this codebase will need to go under report but accessed via the ANC module 

**New structure**
**`/anc`** _- anything ANC specific will build CHW app for LMH and mADX_
`/eir` _- anything EIR specific will build Covax app_
`/engine`
` |--- /register`
**` |--- /report`**
` |--- /task`
` |--- /search`

**Explanation**
This implementation is part of the mADX project that aims a generating measure reports for set predefined indicators. The set of patients are primarily ANC clients, the set of observations and details taken are collected in the course of different ANC encounters and the measures for reporting are a set of records that need to be tracked for various ANC Indicators.

Thus, patients will be accessed via the ANC app and then for those being tracked or in use for measure reporting we can run evaluate operation against the related resources i.e. Library resource with CQL etc

**Comment by f-odhiambo** [comment]
@kimigxfoxy The sample app has since been updated with some new changes and the ability to generate measure reports check it out https://github.com/DBCG/CqlEvaluatorSampleApp We may need to upgrade part of the previous work done Also there us a new very loaded bundle [here](https://raw.githubusercontent.com/DBCG/CqlEvaluatorSampleApp/master/app/src/main/assets/resources/content/measure/ANCIND01/ANCIND01-bundle.json)

**Comment by f-odhiambo** [comment]
**Some additional notes**

Understanding the Library logic vs the measure logic 


The Library has the actual CQL embedded, while the Measure is a higher-level construct that gives the elements of the CQL meaning within the context of a quality measure. Figure 3-1 at the link here illustrates the relationship:
https://build.fhir.org/ig/HL7/cqf-measures/measure-conformance.html#population-criteria

The Parameters resources is the return type defined for the $evaluate operation for Libraries. It roughly corresponds to key-value pairs for the CQL expressions that are embedded:
http://build.fhir.org/ig/HL7/cqf-recommendations/OperationDefinition-cpg-library-evaluate.html
What you do with the Parameters resources is context-dependent. In the case of a Measure, the values in the Parameters would map to population criteria. In the context of clinical decision support, it might map to a CDS-hooks card. If you were building a questionnaire app you might show the results to a user. In other words, it's a relatively low-level operation that can be reused under the covers in many different ways. It might be something that's not exposed at all within the Android SDK, depending on the specific requirements. In the demo app, we merely needed it as a first step towards Measure evaluation.

**Comment by f-odhiambo** [comment]
**Some updates**
1. @RaaziaTarique to share with @kimigxfoxy the Patient profile XML
2. @kimigxfoxy You can add a label and button just under the CarePlan section 
![image](https://user-images.githubusercontent.com/4540684/131983354-b5ef01f9-99ab-4f7c-9261-fb8a596d49b9.png)

**Comment by RaaziaTarique** [comment]
@kimigxfoxy  shared ANC details xml with you. Please check slack.

**Comment by kimigxfoxy** [comment]
@f-odhiambo I am waiting for this PR to be approved _https://github.com/opensrp/fhircore/pull/515_ so that I can pull develop and merge to my branch and the button from there. @RaaziaTarique Thanks

**Comment by kimigxfoxy** [comment]
@f-odhiambo Test patient with anemia https://fhir.labs.smartregister.org/fhir/Patient/1744/$everything.

**Comment by f-odhiambo** [comment]
Part of the integration to ANC for this POC missing @kimigxfoxy to integrate into a new PR  - Access the evaluation from an ANC patient  profile page 

cc @kimigxfoxy

**Comment by kimigxfoxy** [comment]
Library evaluator class merged using PR https://github.com/opensrp/fhircore/pull/557
Next step is adding information to anc details fragment
